### DISCUSSION TITLE
CompositeActivity Introduction

### DISCUSSION THREAD

**Discussion by Nokecy** [elsa-workflows/elsa-core#1291] [Q&A]
need CompositeActivity Introduction

**Comment by craigfowler** [comment]
Could you elaborate a little on what that means please?  What do you want to be able to do that can't already be done?

**Comment by Nokecy** [comment]
public class AbpParallelUserTaskActivity : CompositeActivity
    {
        [ActivityProperty(Hint = "")]
        public List<object> Items
        {
            get => GetState<List<object>>();
            set => SetState(value);
        }

        public override void Build(ICompositeActivityBuilder activityBuilder)
        {
            activityBuilder.ParallelForEach(Items, iterate =>
            {
                iterate
                        .Then<UserTask>(activity =>
                        {

                        })
                        ;
            });
        }
    }


I want to know the number of tasks completed in parallel @craigfowler  can you help me?

**Comment by craigfowler** [comment]
Thanks for this @Nokecy .  I think I understand a little more now.  For a start I assume you are talking about **Elsa 2.0** - because you're asking about `CompositeActivity`?

When you say "the number of tasks completed":

* Do you mean like "the progress of iterating those tasks at a moment in time".  Perhaps you are trying to write a progress indicator so you can say "4 out of 20 are done"
* Do you mean like "the total number _which were actually completed_, after it's finished".  Perhaps you are trying to write a reporting mechanism that says "All done.  There were 18 tasks completed and 2 failed to complete".
* Is it good enough if you were able to refer back to the original `Items` collection from the activity?  That might not technically mean _"how many were actually done"_ but means _"how many you meant to do"_ instead.

**Comment by Nokecy** [comment]
public class CustomizedUserTaskActivity : Activity
    {
        private readonly IContentSerializer _serializer;

        [ActivityProperty(Type = ActivityPropertyTypes.List, Hint = "Provide a list of available actions")]
        public ICollection<string> Actions { get; set; } = new List<string>();

        [ActivityProperty(Label = "Assign User", Hint = "")]
        public string Assign { get; set; } = string.Empty;

        public CustomizedUserTaskActivity(IContentSerializer serializer)
        {
            _serializer = serializer;
        }

        protected override bool OnCanExecute(ActivityExecutionContext context)
        {
            var userAction = GetUserAction(context);

            return Actions.Contains(userAction, StringComparer.OrdinalIgnoreCase);
        }

        protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
        {
            return Suspend();
        }

        protected override IActivityExecutionResult OnResume(ActivityExecutionContext context)
        {
            var userAction = GetUserAction(context);
            return Outcome(userAction, userAction);
        }

        private string GetUserAction(ActivityExecutionContext context)
        {
            return (string)context.Input!;
        }
    }

1、I want to create a task item for assigned users ，May I ask where should I create?
2、If multiple users are assigned and executed in parallel, how to determine the activity is complete

Should I use CompositeActivity or custom Activity?

**Reply by sfmskywalker** [reply]
Hi @Nokecy,

A Composite Activity is like a mini-workflow: it contains activities and connections between them.

Your questions about task items for assigned users and where to create them is a little bit too broad to answer properly.

**Comment by MaddyDecoder** [comment]
is there is any simple workflow realtime example for composite activity . advt and disadvt so that while creating i can know whent o use and when to not ?
